site/content/blog/2019-01-20-submitting-great-prs.md

```diff
@@ -0,0 +1,76 @@
+
+### 3) Clean up your history.
+
+Everyone has their own development style. Mine usually consists of coding a small part of the overall problem, typing:
+
+```sh
+git commit -m wip
+```
+
+and repeating this process often. Commit early and commit often! It's just easier to undo changes this way while you're actively developing.
+
+But a reviewer doesn't need to see that commit history. As a reviewer myself, I'd want to see one of two things. Either:
+
+1. A single squashed commit with all of your changes.
+
+2. Multiple commits breaking up your changes into logical groupings.
+
+Put that variable name fix in its own commit, commit your feature change on top of it, and add a note to your PR's description. Your reviewers can then use the commits tab in GitHub to review those changes independently.
+
+Maybe you're the kind of developer that writes a pristine commit history the first time, but most of us rely on `git rebase`. If you aren't familiar with rebasing, it might seem terrifying or magical, but it doesn't need to be: play with it, master it, and use it often. If you need help getting started here, check out Atlassian's excellent [tutorial on git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
+
+Keep in mind though that rebasing is something to do before you submit your pull request, not after, and this leads to my last point:
+
+### 4) Avoid Force Pushing
+
+Good git etiquette is to avoid force pushing any branch someone else is actively using. A pull request is a branch just like any other, so once you've started that pull request, everything that you push upstream should just be a regular or merge commit. In particular, when merge conflicts arise with `master`, just:
+
+```sh
+git fetch && git merge origin/master
+```
+
+Resolve the conflicts and push up the resulting merge commit. Resist the urge to rebase anything you've already pushed to the pull request.
+
+Why? When you rebase, you are effectively replaying a commit history on top of another commit. But in a pull request, a reviewer wants to see a linear history of your changes from when they started reviewing. Any merge conflicts resolved during the rebase operation are "hidden" inside the original commits instead of appearing as a new commit. A diligent reviewer must then start from the beginning instead of just examining the latest changes. Exacerbating all of this is that GitHub tends to lose or detach comment threads across a rebase, making it difficult to regain context from an earlier discussion.
+
+Note that other open source projects may merge your commit history instead of squashing it as we do here at Mattermost. In that case, I'd recommend checking with the project maintainers and possibly rebasing just before merging to ensure a clean history afterwards.
```

Comment:
RE: "_rebasing just before merging_" - I disagree but may be a little confused here. From what I understand if I have a lengthy and long-lived PR, with a lot of merge commits in it, rebasing before merge is prone to a lot of issues, and would require another review cycle. Thus we hear "rebase often" cited as a best practice. Maybe "_...checking with the project maintainers ~~and possibly rebasing just before merging to ensure a clean history afterwards~~ for the best practice in their projects._"

Reply:
Agree that we should just defer to the project maintainers here -- I think I tried too hard to anticipate the flow in other projects.